### PR TITLE
[#3]Feature/add user defined options.

### PR DESCRIPTION
**Pull request by tamerharsh**
**Description:-**
This feature adds the user defined options.

**Message:-**
Please review.

### CODE REVIEW

File: commandlineparser/include/Parser.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+#include "Options.h"
+
+#ifndef PARSER_H
+#define PARSER_H
+
+namespace cmdParser
+{
+	class Parser 
+	{
+		std::map<std::string ,opts::Options*> command_list;
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: always  include  class members at the bottom of the class

Comment (inbangsa) [inline commandlineparser/include/Parser.h]: better to create std::map<std::string,std::shared_ptr<opts::Options>> command_list

Comment (inbangsa) [inline commandlineparser/include/Parser.h]: please create a alias for map declaretype 

using commandList = std::map<std::string ,opts::Options*>

and then you can use commandList  instead std::map<std::string ,opts::Options*>

File: commandlineparser/include/Parser.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+namespace cmdParser
+{
+	class Parser 
+	{
+		std::map<std::string ,opts::Options*> command_list; 
+		public:
+		void add_options(std::string short_command ,std::string long_command,std::string short_description, std::string long_description);
+		
+    };
+};
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h:33]: destructor missing for the class.. big memory leak as you are creating naked pointer map.. where is the deletion of this map

File: commandlineparser/src/Parser.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#include "Parser.h"
+
+void cmdParser::Parser::add_options(std::string short_command ,std::string long_command,std::string short_description, std::string long_description)
+{	
+	opts::Options *obj=new opts::Options(short_command,long_command,short_description,long_description);
+	
+	if(!(obj->get_option_short_command().empty()))
+	{	
+		command_list.insert(std::pair<std::string,opts::Options*>(obj->get_option_short_command(),obj));
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp]: inside class decleartion do 

using pair_string_option= std::pair<std::string,opts::Options*>;

File: commandlineparser/src/Parser.cpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	}	
+	else if(!(obj->get_option_long_command().empty()))
+	{
+		command_list.insert(std::pair<std::string,opts::Options*>(obj->get_option_long_command(),obj));
+		
+	}
+	else
+	{
+		std::cout<<"Enter a valid short or long command or both";
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp]: If this is not a valid condition throw and exception.

File: commandlineparser/src/Options.cpp [hunk 5]
```diff
@@ -1,25 +1,47 @@
-#include"Options.h" 
+#include "Options.h" 
 
-Options::Options(std::string short_command, std::string long_command,
+opts::Options::Options(std::string short_command, std::string long_command,
```
Comment (inbangsa) [inline commandlineparser/src/Options.cpp]: why two namespaces cmd_parser and opts

**Review by inbangsa — CHANGES_REQUESTED** [review]

File: commandlineparser/include/Parser.h (hunk 1 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: done

File: commandlineparser/src/Options.cpp [hunk 6]
```diff
@@ -1,46 +1,41 @@
...
- { 
-	 set_option_long_command(long_command);
- }
- 
-void opts::Options::set_option_short_command(std::string short_command)
+
+void cmdParser::Options::set_option_short_command(std::string short_command)
 {
-	if(!short_command.empty()){this->short_command='-'+short_command;}
+	if(!short_command.empty())
```
Comment (inbangsa) [inline commandlineparser/src/Options.cpp]: Alwways use {} even for a single statement.

File: commandlineparser/src/Parser.cpp [hunk 7]
```diff
@@ -1,20 +1,31 @@
...
+{
+	std::shared_ptr<cmdParser::Options> obj = std::make_shared<cmdParser::Options>(short_command, long_command, short_description, long_description);
 	
-	if(!(obj->get_option_short_command().empty()))
-	{	
-		command_list.insert(std::pair<std::string,opts::Options*>(obj->get_option_short_command(),obj));
-	}	
-	else if(!(obj->get_option_long_command().empty()))
+	//if both  short and command a present make both  of them individually as key.
+
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:6]: delete this space

File: commandlineparser/src/Parser.cpp [hunk 8]
```diff
@@ -1,20 +1,31 @@
...
+	if ((!(obj->get_option_short_command().empty())) && (!(obj->get_option_long_command().empty())))
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp]: Not needed at all

File: commandlineparser/src/Parser.cpp [hunk 9]
```diff
@@ -1,20 +1,31 @@
...
+		command_list.insert(KeyValue(obj->get_option_short_command(), obj));		
+	}
+	//if only long command is present, make long command as key.
+
+	else if (!(obj->get_option_long_command().empty()))
 	{
-		command_list.insert(std::pair<std::string,opts::Options*>(obj->get_option_long_command(),obj));
-		
+		command_list.insert(KeyValue(obj->get_option_long_command(), obj));
 	}
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:20]: rewrite complete logic make it more optimize

File: commandlineparser/src/Parser.cpp [hunk 10]
```diff
@@ -1,20 +1,31 @@
...
 	else
 	{
-		std::cout<<"Enter a valid short or long command or both";
+		throw std::runtime_error("Enter a valid short or long command or both.");
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp]: send a normal exception

File: sample/sample_user.cpp [hunk 11]
```diff
@@ -1,9 +1,18 @@
...
+#include "Parser.h"
+int  main(int argc, char*argv[])
 {
+	cmdParser::Parser obj;
+	try
+	{
+		obj.add_options("", "port", "short description ", "very very long description");
+		obj.add_options("cp", "copy", "short description ", "very very long description");
+		obj.add_options("IP", "", "short description ", "very very long description");
+		obj.add_options("", "", "short description ", "very very long description");
```
Comment (inbangsa) [inline sample/sample_user.cpp]: Donot put exceptoions checks conditions in example.. this leads to the Topic that we need testframework and testcases for the Library

**Review by inbangsa — CHANGES_REQUESTED** [review]

**Review by inbangsa — APPROVED** [review]

**Pull request merged**